### DISCUSSION TITLE
Please don't remove createContext

### DISCUSSION THREAD

**Discussion by arvinxx** [pmndrs/zustand#1276] [General] [answered]
I notice that https://github.com/pmndrs/zustand/blob/main/docs/previous-versions/zustand-v3-create-context.md says:

> Note: This function will likely be deprecated in v4 and removed in v5.

I think this createContext is more usable that react's createContext. Please don't deprecated or remove it in the further version.

### Reason why this createContext is more usable

ths most useful situation is refactor a app to a component.

Here's progress:

At first I create a App without context :

```ts
// store.ts
import create from "zustand";

export const useStore =  create((set) => ({
    bears: 0,
    increasePopulation: () => set((state) => ({ bears: state.bears + 1 })),
    removeAllBears: () => set({ bears: 0 })
  }));
```

compoents use `useStore` to consume store
```tsx
// Component.ts

import {useStore} from './store'

const ButtonChild = () => {
  const state = useStore();
  return (
    <div>
      {state.bears}
      <button
        onClick={() => {
          state.increasePopulation();
        }}
      >
        +
      </button>
    </div>
  );
};
export default ButtonChild
```

But oneday, I need to convert this app to a component(like [react-flow](https://reactflow.dev/)) to reuse in another app.

with zustand createContext, what I need to do is just wrapper the App with a zustand's createContext, and don't need to refactor any code in children components. It become a component, can be used in any other app.

```diff
// store.ts
import create from "zustand";

+ const createStore = ()=> create((set) => ({
- export const useStore =  create((set) => ({
    bears: 0,
    increasePopulation: () => set((state) => ({ bears: state.bears + 1 })),
    removeAllBears: () => set({ bears: 0 })
  }));

+ const { Provider, useStore } = createContext();

+ export { Provider, useStore ,  createStore }
```

```tsx
// Wrapper.tsx

import {Provider,  createStore } from './store'

const  Wrapper =()=>{
  return (
    <Provider createStore={createStore}>
      <ButtonChild />
    </Provider>
  );
}
```

But what if use it with react's context? 

1. change `create` usage to `createStore` usage
2. Add wrapper with react context.
3. change all children components usage to new useStore. If I have 20 components, I need to rewrite 20 times.

Compare to zustand 's context usage, It's real bad.

So please keep the zustand's createContext usage in the furture. And it's a more recommended method in my mind.

**Comment by dai-shi** [comment, marked as answer]
The point is you can create v3 compatible createContext with v4 new APIs.
We want to make the library minimalistic, and thus considering deprecation.

Your feedback is important and I agree that there are some use cases. I wonder if anyone would be interested in creating a 3rd-party package, if it's valuable.

**Reply by arvinxx** [reply]
> I wonder if anyone would be interested in creating a 3rd-party package, if it's valuable.

Actual when I use zustand, I have considered `zustand/middleware` 、 `zsutand/context` as enhancement (3rd-party package) to `zustand` package.  Although they are in the same package.

> We want to make the library minimalistic, and thus considering deprecation.

I agree it's a trade off. But  I think where to consider it with real world situation.I love zustand's reason is that it cover most usage I need in my production, like selector、devtools、work out of react and so on. It's DX is great! 

I believe convert app to component is also a important use case, and it's also a good pratice in progressive state management in zustand usage. 


> The point is you can create v3 compatible createContext with v4 new APIs.

As a user, actually I don't know how to `create v3 compatible createContext with v4 new APIs`, and actually most users don't care it too. If zustand consider to cover most state management case, I really suggest to keep `createContext` usage. 

If think in this way , there will be lots of similar thoughts that leads to use bare react `useState`.

I more like to find a best practice that can work in any situation.

**Reply by dai-shi** [reply]
> If zustand consider to cover most state management case

I see some gap here. Our understanding is that the major usage of zustand is without context, because it's the selling point and many people like it.
The context usage would be minor compared to it. 

> find a best practice

I feel like this is a documentation issue.

---

From the different aspect, I would like to discourage `zustand/context`. It unnecessarily hides some internals. In v3, it was unavoidable to misusing hooks, which doesn't apply to v4 new API. That's the major reason for deprecation: I didn't want to add it in v3 in the first place.

So, my question is how much is your pain with v4. Here's the diff.

```diff
// store.ts
+ import { createContext, useContext } from "react";
- import create from "zustand";
- import createContext from "zustand/context";
+ import { createStore, useStore } from "zustand";

- const useStore = create((set) => ({
+ const store =  createStore((set) => ({
    bears: 0,
    increasePopulation: () => set((state) => ({ bears: state.bears + 1 })),
    removeAllBears: () => set({ bears: 0 })
  }));

+ const MyContext = createContext()

+ export const Provider = ({ children }) = <MyContext.Provider value={store}>{children}</MyContext.Provider>;

+ export const useMyStore = (selector) => useStore(useContext(MyContext), selector);
```

**Reply by arvinxx** [reply]
>  Our understanding is that the major usage of zustand is without context, because it's the selling point and many people like it.

Yes, I agree it that context-less is really great,ant I use it in all my app. And then I start to think,is there any method to use it in component development?  because I want to use all great features in zustand to develop component library( like [react-flow](https://github.com/wbkd/react-flow), which using zustand too  and win the best react component in 2022).

Then I find that if I want to use zustand in component development,I need to use context  to make single instance store to mutli-instance store. And then zustand can cover almost every store management in real production world.

That's why I said that don't remove createContext. After seeing your code, I think the method can cover the component usage too. (By the way ,how to create the `useStoreApi` hook ? )

And please add this to document too. I think it's useful for other developers.

**Reply by dai-shi** [reply]
> (By the way ,how to create the useStoreApi hook ? )

It's as simple as this:
```js
export const useMyStoreApi = () => useContext(MyContext);
```

**Reply by dai-shi** [reply]
> And please add this to document too. I think it's useful for other developers.

Would you like to work on it? Otherwise, @dbritto-dev might be able to take it?

Here are some other pointers:
- https://github.com/pmndrs/zustand/releases/tag/v4.0.0
- https://github.com/pmndrs/zustand/discussions/1180


https://github.com/pmndrs/zustand/discussions/1180#discussioncomment-3443817 Oh, @Cuppachino is still working on it?
We have a new docs folder structure, so `./docs/zustand-v4-context-usage.md` should be in some folder. cc @chrisk-7777

**Reply by arvinxx** [reply]
> Would you like to work on it? Otherwise, @dbritto-dev might be able to take it?

Yes , I'm glad to add this part to it. Please tell me where should I add to.

**Reply by dai-shi** [reply]
Maybe `./docs/guides/zustand-v4-context-usage.md`. We can revisit the filename and the folder later.
I want it to include what's discussed in #1180.

**Reply by divmgl** [reply]
There's something I'm not following here. I haven't yet found a solution for these two `createContext` use cases:

1. Creating an instance of a Zustand store that is uniquely shared between multiple components. Is this a use-case that Zustand will stop supporting in v5? We find it immensely useful being able to both A) avoid using `React.Context` and B) use Zustand selectors to selectively render components. Leaning on `React.Context` is a surefire way to degrade performance in our app.
2. `createContext` helps us initialize the store so that values are not `undefined`. I have not yet found a way to do this that is type-safe with Zustand, and the `createStore` pattern with `createContext` is the only solution we've found.

We'd love to keep using Zustand. How do the new APIs solve for these use cases? If we don't have a solution for this, we'll be stuck perpetually on v4 because we don't have great alternatives to solve those two issues.

**Reply by dai-shi** [reply]
Have you seen the migration guide? https://github.com/pmndrs/zustand/discussions/1180#discussioncomment-3354713

**Reply by divmgl** [reply]
Let me take a look and see if that answers both of my questions, thanks for pointing me in the right direction.

**Reply by dai-shi** [reply]
You can also reach out a third-party library. https://github.com/charkour/zustand-di

**Reply by josegoval** [reply]
I think in general context usage is extremely useful to create Zustand store instances per component level, and as such, It should be considered a core aspect of Zustand.

> In Zustand we should be able to create global stores (the most common usage) and store instances.

Store instances combine really well with react preventing prop drilling and achieving clear organization within project, among other benefit commented by arvinxx.

**Reply by dai-shi** [reply]
The new docs by @dbritto-dev should address the context usage.

**Comment by arvinxx** [comment]
I just released a library with v3.5 context : https://github.com/arvinxx/zustand-utils. Maybe I can handle with some migration documents context releated. @dai-shi

**Reply by dai-shi** [reply]
Good to see you use `useStore` hook. You may not need useMemo in your useStoreApi.

**Reply by arvinxx** [reply]
> Good to see you use `useStore` hook. You may not need useMemo in your useStoreApi.

improved

**Comment by ollwenjones** [comment]
I fail to understand how moving to react context as in the migration guide allows us to have multiple instances of the same store, with some state passed from a parent component, as before, because these examples are creating the store at the module level. If we create a store instance at the module level (singleton) we can't have more than one instance of the same store for different components / component instances.

TypeScript also says `createContext` requires at least 1 argument, fwiw

**Reply by divmgl** [reply]
This is the same problem we've had with no solution aside from spinning up our own custom solution on top of Zustand using useMemo. It's a shame.

**Reply by dbritto-dev** [reply]
Hey peeps I don't see why you can't have more than one instance of the same store for different components. You can even have a factory method in order to create the same store multiple times and store them in a map/set data structure. Regarding TS you can use `createContext<Store | null>(null)` or `createContext<Store>(null as never)`

**Reply by dai-shi** [reply]
> If we create a store instance at the module level (singleton)

If you create a store at the module level, there's not much point of using React Context. Do docs misguide?

> using useMemo

It works, but for unstable props I'd use `use-constant` (or `useRef`).

For convenience, check out https://github.com/charkour/zustand-di and https://github.com/arvinxx/zustand-utils.

**Reply by ollwenjones** [reply]
what I did before, which was "clever" but not "correct" was to wrap store creation in a hook that stored store instances in a map by ID, so it would create only on the first try. using zustand/context seemed more "correct" which others did elsewhere in our codebase (because why reuse a pattern invented by someone in your own company :trollface: ?) Maybe I shouldn't worry about "correct" if it works 😄 

@dai-shi thanks for responding. Maybe this doc only misguieded me, but it really looks like in [this example](https://docs.pmnd.rs/zustand/previous-versions/zustand-v3-create-context#migration), everything is at the module level. 

If you call
```js
+ const store = createStore((set) => ({ ... }));
  ```
 in a component won't it create a new store at every render? This is what seems like is happening for me.
 
 ```js
 export const Provider = ({ children }) => <MyContext.Provider value={store}>{children}</MyContext.Provider>;
 ```
 
 sure seems to be module level. I can't think why you would need context in that case... 

And if you use a ref and something like an effect to create the store in the parent one time, you still have to deal with it being `undefined` on the intial render of the child components when you call `useStore(useContext(MyContext), selector)`

What am I missing? This migration doc makes no sense to me. I can solve it my old "clever" but not quite correct way, sure, but if there's a right way to do it with context that is less "in the component lifecycle" (since that's bad right?). I would like to understand.

**Reply by dai-shi** [reply]
> https://docs.pmnd.rs/zustand/previous-versions/zustand-v3-create-context#migration

Okay, I now see this migration guide is incomplete or misguiding. It was trying to show the simple diff, but it's not even "correct".

For migration, I think the thrid-party libs make sense.

So, let me write the "correct" usage in v4. (btw, `createContext` in v3 works, but it's a workaround or a hack. That's why we wanted to deprecate.)

```jsx
import { createContext, useContext, useRef } from 'react'
import { createStore, useStore } from 'zustand'

const StoreContext = createContext(null) // TS requires `null`, but JS doesn't.

const StoreProvider = ({ children, ...otherPropsIfYouWant }) => {
  const storeRef = useRef()
  if (!storeRef.current) {
    storeRef.current = createStore((set) => ...)
  }
  return <StoreContext.Provider value={storeRef.current}>{children}</StoreContext.Provider>
}

const useStoreInContext = (selector) => {
  const store = useContext(StoreContext)
  if (!store) { throw new Error("Missing StoreProvider") }
  return useStore(store, selector)
}
```

**Reply by ollwenjones** [reply]
That's great! Thanks for being so responsive! Now that I see it, it seems obvious, haha. Maybe I would have thought of it if I hadn't hit the problem late on a Friday

**Reply by dai-shi** [reply]
#2039 to fix the doc.

**Reply by norama** [reply]
I am trying to use this approach but having some fatal problem upon reloading the page. But before digging into this, just a quick remark:
```javascript
 const storeRef = useRef()
  if (!storeRef.current) {
    storeRef.current = createStore((set) => ...)
  }
```
is in my opinion equivalent to
```javascript
 const storeRef = useRef(createStore((set) => ...))
```
or am I missing sg here?

**Reply by norama** [reply]
So now the fatal problem: we have a big store with quite a lot of actions going on and at some point I realized that the store is not initialized upon page reload. After a little bit of debugging it turned out that for some reason there were 2 stored created and while updates referred to the old one, the UI used the second one which was empty. Things got sorted out upon subsequent store updates but this first reload always had this problem. Then I came up with the following solution: just put the store into React state instead of React ref everything is fine, no duplicate stores are created:

```javascript
const StoreProvider = ({ children, ...otherPropsIfYouWant }) => {
   const [store, setStore] = useState(null)

  useEffect(() => {
    setStore(createStore((set) => ...))

    return () => setStore(null)
  }, [])

  return store ? (
     <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
  ) : null
}
```

Also, it seems to me logical to require a re-render upon store change.
What do you think?

**Reply by norama** [reply]
But in another use case even that fails, there are 2 stores again and the same problem arises.
Maybe the single reliable solution is using global store without context as you have in the intro:
https://docs.pmnd.rs/zustand/getting-started/introduction

**Reply by dai-shi** [reply]
```js
  const storeRef = useRef(createStore((set) => ...))
```
is equivalent to
```js
  const initialStore = createStore((set) => ...)
  const storeRef = useRef(initialStore)
```

> there were 2 stored created

Can you create a minimal reproduction with https://codesandbox.io/s/react-new and open a new discussion?

> just put the store into React state

That is fine from the library usage perspective.

**Reply by norama** [reply]
You can ignore this, it was caused by accidental forced remount in our code.
Anyway, I have also learned why
```
const storeRef = useRef()
  if (!storeRef.current) {
    storeRef.current = createStore((set) => ...)
  }
```
is better than setting 
```
const storeRef = useRef(createStore((set) => ...))
```
right away: this latter variant calls `createStore((set) => ...)` upon each render with no purpose, the new store will not be used.
So it will work but will generate extra stores that are not used.

**Reply by xinaesthete** [reply]
I have a different issue... I have some code (third party code I'm trying to refactor/adapt) that uses `useMyStore.setState()` from event listeners etc.

So if I have something like (excuse my condensed example code):

```js
const useMyStore = selector => {
  return useStore(useContext(StoreContext), selector)
}
```

Then `useMyStore.setState` is undefined... but if I try

```js
function useMyStore(selector) {
  return useStore(useContext(StoreContext), selector)
}
useMyStore.setState = s => {
  const store = useContext(StoreContext) //uh-oh
  store.setState(s)
}
```

Then it won't be able to work because of rules of hooks... Perhaps it would be preferable to replace any such instances with more specific store actions - which is doable, but then this refactor will need to have a higher surface area and diverge from the original more than I would ideally prefer.

Excuse me if I missed something.

**Reply by dbritto-dev** [reply]
@xinaesthete that's expected since the context is undefined until the component is mounted if you want re-use an store outside of react you should use `createStore`

**Reply by marcebdev** [reply]
From what I can find I'm seeing the same issue as @norama in regards to React.StrictMode? Is this something the example just doesn't account for?

**Reply by dbritto-dev** [reply]
@marcebdev is the same issue with `useState` -> `useState(() => createStore())` instead of `useState(createStore())` the `useState(createStore())` would create multiple stores

**Comment by xinaesthete** [comment]
> @xinaesthete that's expected since the context is undefined until the component is mounted if you want re-use an store outside of react you should use `createStore`

Perhaps I should've kept a bit more code for context - I was using `createStore` in what I believe to be a similar pattern [this documentation](https://docs.pmnd.rs/zustand/previous-versions/zustand-v3-create-context#migration) which by my reading I thought was the most canonical...

Indeed, now that I review a bit more of the docs, I see https://docs.pmnd.rs/zustand/guides/practice-with-no-store-actions

> this pattern doesn't offer any downsides

... but it seems to offer something of a downside when it comes to adapting it to the context API.

I could maybe benefit from re-reading some of this discussion, as well as reviewing the utility libraries mentioned... heading out now, but FWIW [here is a code sandbox with a simple store](https://codesandbox.io/p/sandbox/zustand-context-ts-y5cpyh?file=%2Fsrc%2Fzustand.tsx%3A17%2C1) (and lots of boilerplate for types etc).

I keep a `Map` of stores, and have another hook `useStoreGetter` that returns a function that when called (later, outside react component render) will retrieve the given store. It's then possible to use `setState` and `getState`... it should remove the reference when the provider is unmounted... I don't think there's anything terribly wrong with it, but I'm not sure how much I like it, sure it can be improved. JS version is a bit more brief:

```jsx
import { createContext, useContext, useEffect, useRef } from "react";
import { createStore } from "zustand";
import { useStoreWithEqualityFn } from "zustand/traditional";

const StoreContext = createContext(null);
const stores = new Map();
export const StoreProvider = ({ children, id }) => {
  const storeRef = useRef();
  if (!storeRef.current) {
    console.log("init store", id);
    storeRef.current = createStore(() => ({
      count: 0,
    }));
  }
  useEffect(() => {
    if (storeRef.current) stores.set(id, storeRef.current);
    return () => {
      stores.delete(id);
    };
  }, [storeRef.current, id]);
  return (
    <StoreKeyContext.Provider value={id}>
      <StoreContext.Provider value={storeRef.current}>
        {children}
      </StoreContext.Provider>
    </StoreKeyContext.Provider>
  );
};

export function useStoreInContext(selector) {
  const store = useContext(StoreContext);
  if (!store) {
    throw new Error("Missing StoreProvider");
  }
  //@deprecated...
  //return useStore(store, selector);
  return useStoreWithEqualityFn(store, selector);
}
/**
 * const storeGetter = useStoreGetter()
 * useEffect(() => {
 *   storeGetter().setState(...)
 * })
 */
export function useStoreGetter() {
  const k = useContext(StoreKeyContext);
  return () => stores.get(k);
}
```

Any thoughts welcomed (including on the type stuff, which it may be possible to make a bit less involved on the provider side with some thought...)

**Reply by xinaesthete** [reply]
I just realised I'd mis-pasted the sandbox link above, should have been https://codesandbox.io/p/sandbox/zustand-context-ts-y5cpyh?file=%2Fsrc%2Fzustand.tsx%3A17%2C1 (now edited in the previous comment as well)

**Reply by RinNguyens** [reply]
I have added Immer middleware insight createStore. It's works. But I don't know, It good for context when call any components.
![immer](https://github.com/pmndrs/zustand/assets/55022071/78807404-2fa1-4788-9bc4-927a0ebd3d25)

**Comment by xinaesthete** [comment]
> I have added Immer middleware insight createStore. It's works. But I don't know, It good for context when call any components.

It might be better if you include the code as text rather than an image. I'm not actively using Immer at the moment and don't have any immediate thoughts on whether there are problems with your approach, but may review at some point.
